### PR TITLE
Basic Player Implementation

### PR DESCRIPTION
**Pull request by SamBumgardner**
This PR implements a basic version of the Player class for this Mario-esque platformer.

Major changes in this PR include:
 * Created the Player class
     * Player can move left/right with A/D or the Left/Right arrow keys.
     * Player can jump (when on the ground) by pressing "X" or "/"
          * Note: the Player has no control over their momentum while in the air.
     * Player can run (increasing max speed) by holding "Z" or "."
     * Player is affected by gravity.
 * Added the Player to the PlayState
 * Added an environment tilemap to the PlayState
 * Added collision checks between the Player against the environment tilemap

To test this PR, checkout this branch on your local machine, navigate a command prompt to the directory with this game's project.xml, and run the command `lime test neko`

Please respond with critiques and comments on the code below. There are at least a few things that definitely need fixing/improvement, so see if you can track them down!

### CODE REVIEW

**Review by Cortillaen — COMMENTED** [review]
Lots of comments/questions since I'm still trying to get HaxeFlixel figured out.

### PlayState.hx
* ln23-41:  Is this the standard way of making a level's map?  Also, does this require the entire level be loaded at once?
* ln48:  FlxG.collide(map, player) - I was looking at the documentation for this.  Am I correct in thinking we could group all entities (player, enemies, anything that moves around) together and all blocks, platforms, and any other terrain as a second group and handle collision resolution between the two groups in one call?  For player-enemy/trap collisions, I suppose we will need to build a custom collision handler with FlxG.overlap() and our own function (process who gets hurt, handle death, etc) in place of FlxObject.separate() and call that right after the FlxG.collide() call.  Possibly separate calls for different groups the player can run into (traps might be separate since the player can't damage them, whereas collisions with enemies could damage either entity), though the documentation on collide() and overlap() sounds like minimizing the number of calls is thought to be best for performance.

### Player.hx
* ln33-34:  I don't quite follow the right side of this operation.  It's an unsafe cast of FlxG.state (which is a FlxState object), and then accessing its GRAVITY property.  I assume this is supposed to cast FlxG.state to the PlayState class defined in PlayState.hx (which you give a GRAVITY parameter), but how does it know to do this?
* ln35:  Why is maxVelocity.y set to acceleration.y?  Also, is it efficient to simulate gravity via a constant acceleration instead of only applying that acceleration when not on the ground (mostly, does this cause constant corrections via collision-handling)?
* ln59&72:  Might be worthwhile changing the comments to indicate basis on key-presses rather than movement.  Took me a minute to figure out hoizontalMove stores input, not current state.
* ln60,73,86:  Is there a particular reason the latter two are not else if statements?
* ln135:  Can this distinguish if the object below the Player is a block, enemy, or anything else?

**Comment by SamBumgardner** [comment]
Thanks for the questions!

I'll likely have time to answer them later (possibly much later) tonight. If for some reason I don't, I'll get responses posted first thing tomorrow morning.

**Comment by SamBumgardner** [comment]
I'll do my best to answer each of your questions in the order you asked them in.

## PlayState.hx

* ln23-41: Is this the standard way of making a level's map? Also, does this require the entire level be loaded at once?

Sort of. Tilemaps are pretty standard, but this method of instantiating one isn't necessarily the best.

**Elaboration on using tilemaps (instead of other objects) for the level's environment:** 

You could have each piece of ground be a distinct sprite, but the overhead of creating each of them (and the memory wasted on unused variables) isn't exactly ideal. Instead, we use HaxeFlixel's tilemaps to store a level's static environment in a lean, performance-friendly format. Also, collision detection with a tilemap is generally more efficient than colliding a group of sprite-inheriting variables. 

**Elaboration on how to instantiate tilemaps:**

There are multiple ways to set up tilemaps, some of which have more flexibility (and complexity) than others. For this little demo, instantiating the tilemap from a 1-D array is fine. It's a relatively small chunk of code that gets the job done without much fuss. For a more complex level, we'll look at external tools that give us a nice GUI for level creation/editing.


* ln48: FlxG.collide(map, player) - I was looking at the documentation for this. Am I correct in thinking we could group all entities (player, enemies, anything that moves around) together and all blocks, platforms, and any other terrain as a second group and handle collision resolution between the two groups in one call? 

Yes!

* For player-enemy/trap collisions, I suppose we will need to build a custom collision handler with FlxG.overlap() and our own function (process who gets hurt, handle death, etc) in place of FlxObject.separate() and call that right after the FlxG.collide() call. Possibly separate calls for different groups the player can run into (traps might be separate since the player can't damage them, whereas collisions with enemies could damage either entity), though the documentation on collide() and overlap() sounds like minimizing the number of calls is thought to be best for performance.

I think you have the general idea down. FlxG.overlap() lets us pass in a function that gets called if an overlap event is detected between an object in the first group and an object in the second group. You can think of FlxG.collide() as a special version of FlxG.overlap() that always uses the separate() function as the "function to call if overlap is detected" parameter. 

Also, if we really wanted to cut down on our number of calls to FlxG.overlap(), we could write a sort of  mega-function that in turn determines what particular functions should be called to resolve the collision between whatever two objects are in contact with one another. 

## Player.hx

* ln33-34: I don't quite follow the right side of this operation. It's an unsafe cast of FlxG.state (which is a FlxState object), and then accessing its GRAVITY property. I assume this is supposed to cast FlxG.state to the PlayState class defined in PlayState.hx (which you give a GRAVITY parameter), but how does it know to do this?

When I do `(cast FlxG.state)`, I'm taking `FlxG.state`, which is stored as a variable of the `FlxState` type and telling HaxeFlixel to ignore its typing. At this point, I can attempt to access any property or any function of the casted FlxG.state object, as if it were any type that I want it to be. I could even try something like 
`(cast FlxG.state).jumpUpAndDown()` (a method that the current state doesn't have) and the code would compile without any complaints. The program would crash as soon as I tried to access a property/function that doesn't exist, but it would run up until that point. Since I'm accessing a property that **does** exist in the object stored in FlxG.state, the code does what I intended. 

This is a bit dangerous, though, because if the current state doesn't have a GRAVITY property, then doing the unsafe cast and trying to access GRAVITY will result in a crash every time. 

* ln35: Why is maxVelocity.y set to acceleration.y? Also, is it efficient to simulate gravity via a constant acceleration instead of only applying that acceleration when not on the ground (mostly, does this cause constant corrections via collision-handling)?

I get the feeling that you're right about this being a bit inefficient, but it does simplify some situations, like making sure the player begins falling as soon as they walk off of a platform. You could manually identify that the player is still just touching the ground on every frame (which would save on some calculations for scooting the player back out of the ground). That would be a fair amount of work to put together and test, though, which is why I didn't really worry about it for this basic version of the Player class.

* ln59&72: Might be worthwhile changing the comments to indicate basis on key-presses rather than movement. Took me a minute to figure out hoizontalMove stores input, not current state.

That's a fair point, I'll adjust the comments to make it clear that I'm just narrating the `if` statements.

* ln60,73,86: Is there a particular reason the latter two are not else if statements?

Nope, they definitely could be. I'll take care of that when editing the comments.

* ln135: Can this distinguish if the object below the Player is a block, enemy, or anything else?

The flags used in the `isTouching()` function are set during the call to `separate()` that happens when the PlayState uses `FlxG.collide()` to collide the player against level's tilemap. Basically, if the player collides with (and gets pushed out of) the floor on a particular frame, `isTouching(FlxObject.Down)` will return true for that one frame. This means that any call to `FlxG.overlap()`, which we'll be using for pretty much all player-enemy collisions (or any situation where the two things can overlap) won't set the flags for `isTouching()` by default. If we want to, we can, but it's up to us to decide if a collision should change the player's touching status in any way.


Hopefully that clears up any questions you had. Let me know if anything else comes to mind!

File: source/Player.hx [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+import flixel.FlxObject;
+import flixel.FlxSprite;
+import flixel.system.FlxAssets.FlxGraphicAsset;
+import flixel.util.FlxColor;
+import flixel.input.keyboard.FlxKey;
+
+/**
+ * ...
+ * @author Sam Bumgardner
+ */
```
Comment (SamBumgardner) [inline source/Player.hx:13]: There really should be some sort of documentation here at the top of the file!

File: source/Player.hx [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+		{
+			horizontalMove--;
+		}
+		if (FlxG.keys.anyPressed([FlxKey.RIGHT, FlxKey.D]))
+		{
+			horizontalMove++;
+		}
+		
+		// Determine if attempted to jump
+		var attemptedJump:Bool = FlxG.keys.anyJustPressed([FlxKey.X, FlxKey.BACKSLASH]);
```
Comment (sjf60780) [inline source/Player.hx]: On your commit comment, you said that the player can jump with the '/' key. But in the code you reference the '\' key instead. I would recommend changing the key to the '/' as it is next to the '.' key for going into "maximum overdrive".

File: source/Player.hx [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+		
+		if (isTouching(FlxObject.DOWN))
+		{
+			onGround = true;
+		}
+		
+		
+		if (onGround)
+		{
+			if (attemptedJump)
```
Comment (Dwooll94) [inline source/Player.hx:241]: If you walk off of a platform you can jump while already in the air.

**Review by Dwooll94 — CHANGES_REQUESTED** [review]
Don't allow jumping while already in the air.

File: source/Player.hx (hunk 2 above)
Comment (SamBumgardner) [reply source/Player.hx]: An excellent suggestion, I'll implement that right away!

File: source/Player.hx (hunk 3 above)
Comment (SamBumgardner) [reply source/Player.hx:241]: Nice catch, I'll go fix that now.

File: source/Player.hx [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+		}
+		else
+		{
+			airMovement(isRunning, horizontalMove, elapsed);
+		}
+		
+		super.update(elapsed);
+	}
+	
+}
```
Comment (CheezitBandit) [inline source/Player.hx]: Github is sad that you don't have a newline at the end of this file. :(

**Review by CheezitBandit — CHANGES_REQUESTED** [review]

**Review by Dwooll94 — APPROVED** [review]

**Review by CheezitBandit — APPROVED** [review]

**Review by KeithCissell — APPROVED** [review]

**Pull request merged**